### DISCUSSION TITLE
Combining PINN with FVM

### DISCUSSION THREAD

**Discussion by chang-change** [lululxvi/deepxde#861] [Q&A] [answered]
Hi, Dr. Lu Lu!
Inspired by [this paper](https://www.sciencedirect.com/science/article/pii/S0045782522001906), I am trying to combine PINN with FVM, but I don't know how to get the values around the collocation points. For example, $u_{e}$ , $u_{w}$ , $v_{n}$ , $v_{s}$.
![image](https://user-images.githubusercontent.com/45302365/186053280-7ca7cb60-958e-4e14-a2a4-7793127baa90.png)
Is it possible to get values around collocation points in DeepXDE other than using Taylor series approximation?
Thanks!

**Comment by lululxvi** [comment]
You could compute for any point, as PINN is a network. To implement it in DeepXDE, it is possible and there might be multiple ways. But it is tricky, and I am not sure what the best way is.

**Reply by chang-change** [reply]
Thanks for your reply. After the neural network is trained, I can get any value of points by using `model.predict()`. However, I want to get the values around the collocation points during the training of the neural network. I don't know how to implement it, could you give me some advice? Do I need to modify the underlying code of DeepXDE? Do I need to design a specialized neural network structure? Thanks again.

**Reply by forxltk** [reply]
You could define a special network structure by `apply_output_transform`.  Then just pass the coordinates of those points around the collocation points.

It should be something like:
```
def modify_output(X, Y):
    x = X[:, 0:1]
    x_star = sample_function(x)  # sample_function provides the coordinates of your interest points

    Define your model

    y = your_model(x)
    y_star = your_model(x_star)

    output = tf.concat([y, y_star], axis=1)
    return output

net.apply_output_transform(modify_output)
```

**Reply by chang-change** [reply]
@forxltk Thanks for your reply. I have a question. 
Since the input to the neural network is `x` and not `x_star`, how can I get the derivative of `y_star` with respect to `x_star` and add it to the PDE loss function?
```python 
def pde(X,Y):
    x = X[:,0:1]
    x_star = ???
    y = Y[:,0:1]
    y_star = Y[:,1:2]  
    dy_x =  dde.grad.jacobian(y,x,i =0,j = 0)
    dystar_xstar = ???
```
I don't think `x_star` is included in `X`. Thanks again.

**Reply by forxltk** [reply]
It's the same. Compute the derivative in `modify_output`, then make it as an additional output.

**Reply by chang-change** [reply]
@forxltk Thank you very much for your prompt reply, I will try it.

**Reply by chang-change** [reply]
@forxltk 
Sorry to bother you again. I want to implement the output of the neural network to be the collocation point and the derivative of the left and right sides of the collocation point with respect to `x_left` and `x_right`, is it possible to write it like this. The code:
```python 
def modify_output(X,Y):
    
    x = X[:,0:1]
    y = X[:,1:2]
    x_right = x + 0.1
    x_left = x -0.1
    y_top = y + 0.1
    y_bottom = y -0.1
    
    
    point_right = tf.concat([x_right,y],axis = 1)
    point_left = tf.concat([x_left,y],axis = 1)
    point_top = tf.concat([x,y_top],axis = 1)
    point_bot = tf.concat([x,y_bottom],axis = 1)
    
    my_model = models.Sequential()
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(1, activation='tanh'))

    center_point_val = my_model(X)
    right_point_val = my_model(point_right)
    left_point_val = my_model(point_left)
    top_point_val = my_model(point_top)
    bot_point_val = my_model(point_bot)
    
    # dcenter_x = dde.grad.jacobian(center_point_val,X,i = 0,j = 0)
    # dcenter_y = dde.grad.jacobian(center_point_val,X,i = 0, j =1)
    dright_x = dde.grad.jacobian(right_point_val,point_right,i = 0,j = 0)
    dleft_x = dde.grad.jacobian(left_point_val,point_left,i = 0,j = 0)
    
    
    final_output = tf.concat([center_point_val, dright_x, dleft_x], axis=1)
   
    return final_output
net.apply_output_transform(modify_output)
```
In addition, I have a question. How can I get many random points in the 2D neighborhood of each collocation point? For example, `red star` is the collocation point, I need to get the coordinates of those blue points. The same sampling method is required for each collocation point.
![Figure_1](https://user-images.githubusercontent.com/45302365/187055676-3f5468b9-7a36-4452-8905-53f4b2262e1e.png )

**Reply by forxltk** [reply]
1. Yes, looks good.
2. Maybe you should define your sample method.

**Reply by chang-change** [reply]
I have defined a sampling function, but when I try to use a `for` loop for each collocation point, the program throws an error and cannot use `for` on tensors.
```python
OperatorNotAllowedInGraphError: iterating over `tf.Tensor` is not allowed in Graph execution. Use Eager execution or decorate this function with @tf.function.
```

**Reply by forxltk** [reply]
I have no experience with it. What about use pytorch backend? https://github.com/lululxvi/deepxde/discussions/789#discussioncomment-3114287

**Reply by chang-change** [reply]
I wrote a sampling function, but I'm not sure if it's correct and if its output is the points around a collocation point or the set of points around all collocation points?
```python 
def sample_func(xmin,xmax,ymin,ymax):
    x_points = tf.random.uniform(tf.shape(xmin),xmin,xmax)
    y_points = tf.random.uniform(tf.shape(ymin),ymin,ymax)
    x_y = tf.concat([x_points,y_points],axis = 1)
    return x_y

def modify_output(X,Y):
    x = X[:,0:1]
    y = X[:,1:2]
    x_right = x + 0.1
    x_left = x -0.1
    y_top = y + 0.1
    y_bottom = y -0.1
    
    xy = sample_func(x_left,x_right,y_bottom,y_top)

    my_model = models.Sequential()
    my_model.add(layers.Dense(64, activation='tanh'))
    .............
    my_model.add(layers.Dense(1, activation='tanh')) 

    source  = my_model(xy)
    source_mean = tf.reduce_mean(source)

    Y = my_model(X)
    final_output = tf.concat([Y, source_mean ], axis=1)
    return final_output 

```
Does `source` represent the values of all sample points around a configuration point? I want to calculate the value of the sampling points around each collocation point and average it. Ultimately, each collocation point should have a corresponding `source_mean`, the average value of all sampling points in the neighborhood of the collocation point.

**Reply by forxltk** [reply]
> Does `source` represent the values of all sample points around a configuration point?

I think it represents all sample points around all configuration points.

But `x_min`, `x_max` ... are (N, 1). Then `sample_func` only samples one point for one configuration point. They should be repeated `num` times.

**Reply by chang-change** [reply]
If I modify `sample_function` to look like this:
```python
def sample_func(xmin,xmax,ymin,ymax,num = 10):
    x_points = tf.random.uniform((len(xmin),num),xmin,xmax)
    y_points = tf.random.uniform((len(ymin),num),ymin,ymax)
    x_y = tf.concat([x_points,y_points],axis = 1)
    return x_y
```
If the output of `sample_func` are `(N,num)`? The `num` represents `num` points for one configuration?

**Reply by forxltk** [reply]
Yes. `x_y` is `(N, 2*num)`. But the input for `my_model` should be `(n, 2)`. You have to reorganize those points by your problem.

**Reply by chang-change** [reply]
When I use `x_points = tf.random.uniform((len(xmin),num),xmin,xmax)`, there is an error:`TypeError: len is not well defined for symbolic Tensors.` How can I make my `sample_func` to enter the coordinates of one sample point at a time?

**Reply by forxltk** [reply]
Use `tf.shape(xmin)[0]`

**Reply by chang-change** [reply]
I modified `sample_func`, and in order to verify the accuracy of the function, my solution field has only one configuration point and no boundary points. However, the following error still occurs:`ValueError: Shape must be rank 2 but is rank 0 for '{{node concat_58}} = ConcatV2[N=4, T=DT_FLOAT, Tidx=DT_INT32](sequential_10/dense_125/Tanh, strided_slice_57, strided_slice_58, Mean_3, concat_58/axis)' with input shapes: [?,1], [?,1], [?,1], [], [].`
```python
def sample_func(xmin,xmax,ymin,ymax,num = 10):
    x_points = tf.random.uniform((num,tf.shape(xmin)[0]),xmin,xmax)
    y_points = tf.random.uniform((num,tf.shape(xmin)[0]),ymin,ymax)
    x_y = tf.concat([x_points,y_points],axis = 1)
    
    return x_y  
```

**Reply by forxltk** [reply]
Don't know where is the problem. May I view all the code?

**Reply by chang-change** [reply]
```python 
def sample_func(xmin,xmax,ymin,ymax,num = 10):

    x_points = tf.random.uniform((num,tf.shape(xmin)[0]),xmin,xmax)
    y_points = tf.random.uniform((num,tf.shape(xmin)[0]),ymin,ymax)
    x_y = tf.concat([x_points,y_points],axis = 1)
    
    return x_y   


def modify_output(X,Y):

    x = X[:,0:1]
    y = X[:,1:2]
    
    x_right = x + 0.1
    x_left = x -0.1
    y_top = y + 0.1
    y_bottom = y -0.1
    
    
    # for i,j,k,z in zip(x_right,x_left,y_top,y_bottom):
    xy = sample_func(x_right,x_left,y_top,y_bottom)

    
    
    point_right = tf.concat([x_right,y],axis = 1)
    point_left = tf.concat([x_left,y],axis = 1)
    point_top = tf.concat([x,y_top],axis = 1)
    point_bot = tf.concat([x,y_bottom],axis = 1)
    
    my_model = models.Sequential()
    # my_model.add(layers.Dense(2, activation='tanh'))
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(1, activation='tanh'))

    center_point_val = my_model(X)
    right_point_val = my_model(point_right)
    left_point_val = my_model(point_left)
    top_point_val = my_model(point_top)
    bot_point_val = my_model(point_bot)
    
    source = my_model(xy)
    # print(source)
    source_mean = tf.reduce_mean(source)
    # print(source_mean)
    
    # dcenter_x = dde.grad.jacobian(center_point_val,X,i = 0,j = 0)
    # dcenter_y = dde.grad.jacobian(center_point_val,X,i = 0, j =1)
    dright_x = dde.grad.jacobian(right_point_val,point_right,i = 0,j = 0)
    dleft_x = dde.grad.jacobian(left_point_val,point_left,i = 0,j = 0)
    
    
    final_output = tf.concat([center_point_val, dright_x, dleft_x,source_mean], axis=1)
   
    return final_output
```

**Reply by forxltk** [reply]
> Yes. `x_y` is `(N, 2*num)`. But the input for `my_model` should be `(n, 2)`. You have to reorganize those points by your problem.

In your code, `x_y` is `(num, 2N)`, which is still inconsistent with `X`.

**Reply by chang-change** [reply]
However, I only collected one point, `N= 1`, in my solution domain.
```python
data = dde.data.PDE(
    geom,
    pde,
    bc,
    num_domain=1,
    
    # num_boundary=400,
    train_distribution = 'Hammersley',
    # solution=func,
    # auxiliary_var_function=source_fun,
    # num_test=1000,
)
```

**Reply by forxltk** [reply]
- Use `source_mean = tf.reduce_mean(source, axis=1, keepdims=True)`. 
- Again, the inputs should be `(n, 2)`, where `2` is `x` and `y`. But your inputs `x_y` for the model is `(num, 2N)`, `N` is the number of sample points. Some errors will be occurred if `N` is not `1`

**Reply by chang-change** [reply]
Thank you. But I don't know how to deal with the shape of the tensor `x_y`, so that it satisfies the input shape of the neural network. Can you offer some suggestions?

**Comment by forxltk** [comment, marked as answer]
I can tell you this idea, however, the code should be done by yourself.
- In `sample_func`, `x_points` should be `(n, num)` that represents `n` sample points with `num` neighbors. 
- Flatten `x_points` so it will be `(x1_1, x1_2, ...x1_num, x2_1, ..., xn_num)`, and the shape is `(n*num, 1)`.
- Concatenate `x_points` and `y_points` as `x_y`. `x_y` should be `(n*num, 2)`
- The output of `my_model` is `(n*num, 1)`. Reorganize the output with shape `(n, num)`
- Compute the mean of your interest dimension. The result should be `(n, 1)` as other tensors in `final_output`

**Reply by chang-change** [reply]
Thank you very much. I'll give it a try. This idea is really great!

**Reply by chang-change** [reply]
```python 
num = 100
def sample_func(xmin,xmax,ymin,ymax,num = num):

    x_points = tf.random.uniform((tf.shape(xmin)[0],num),xmin,xmax)   # x_points_shape = (n,num)
    y_points = tf.random.uniform((tf.shape(ymin)[0],num),ymin,ymax)   # y_points_shape = (n,num)
    
    x_ = tf.reshape(x_points,[tf.shape(xmin)[0]*num,1])    # x_shape = (n*num,1)
    y_ = tf.reshape(y_points,[tf.shape(ymin)[0]*num,1])    # y_shape = (n*num,1)
                    
    x_y = tf.concat([x_,y_],axis = 1)   # x_y_shape  = (n*num,2)
    
    return x_y   

def modify_output(X,Y):
    x = X[:,0:1]
    y = X[:,1:2]
   
    x_right = x + 0.1
    x_left = x -0.1
    y_top = y + 0.1
    y_bottom = y -0.1
    
    xy = sample_func(x_right,x_left,y_top,y_bottom)

    point_right = tf.concat([x_right,y],axis = 1)
    point_left = tf.concat([x_left,y],axis = 1)
    point_top = tf.concat([x,y_top],axis = 1)
    point_bot = tf.concat([x,y_bottom],axis = 1)
    
    my_model = models.Sequential()
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(64, activation='tanh'))
    my_model.add(layers.Dense(1, activation='tanh'))

    center_point_val = my_model(X)
    right_point_val = my_model(point_right)
    left_point_val = my_model(point_left)
    top_point_val = my_model(point_top)
    bot_point_val = my_model(point_bot)
##################################################################    
    source = my_model(xy)  # source_shape = (n*num,1)
    reshape_source = tf.reshape(source,[tf.shape(x)[0],num])   # reshape_source = (n,num)
    source_mean = tf.reduce_mean(reshape_source, axis=1, keepdims=True)  # source_mean_shape = (n,1)
##################################################################    
    dright_x = dde.grad.jacobian(right_point_val,point_right,i = 0,j = 0)
    dleft_x = dde.grad.jacobian(left_point_val,point_left,i = 0,j = 0)
        
    final_output = tf.concat([center_point_val, dright_x, dleft_x,source_mean], axis=1)
   
    return final_output
```
This is my final code. Should I write it this way? Thanks.

**Reply by forxltk** [reply]
Looks good, but still needs to be verified.

**Reply by chang-change** [reply]
With some modifications, the code can run successfully. However, I have a question. If the output of the neural network is four variables, should I design four neural network structures?
 For example, `final_output = tf.concat([output_1, output_2, output_3,output_4], axis=1)`. For each neural network output, there is a different neural network.
```python 
def modify_output(X,Y):
    my_model_1 = models.Sequential()
    ……
    my_model1.add(layers.Dense(1, activation='tanh'))

    my_model_2 = models.Sequential()
    ……
    my_model2.add(layers.Dense(1, activation='tanh'))

    my_model_3 = models.Sequential()
    ……
    my_model3.add(layers.Dense(1, activation='tanh'))

    my_model_4 = models.Sequential()
    ……
    my_model4.add(layers.Dense(1, activation='tanh'))

    # X1,X2,X3,X4 represent different inputs.
    output1 = my_model1(X1)
    output2 = my_model1(X2)
    output3 = my_model1(X3)
    output4 = my_model1(X4)

    final_output = tf.concat([output_1, output_2, output_3,output_4], axis=1)
    return final_output
```
Or just need a neural network with four neurons in output layer? But if this is implemented, how can I map the output variable to the corresponding neuron. Is there something like `component=0,component=1` for this?

**Reply by forxltk** [reply]
I think both are doable. Do all the outputs follow the same mapping? 

>But if this is implemented, how can I map the output variable to the corresponding neuron. Is there something like component=0,component=1 for this?

See DeepONet-FNO paper.

**Reply by chang-change** [reply]
The outputs are like $u,v,w,p$. If the inputs of neural network are same, can I just use one neural network with four neurons in the output layer to represent? The code is following.
```python 
    my_model = models.Sequential()
    ……
    my_model.add(layers.Dense(4, activation='tanh'))

    u = my_model(X)[:,0:1]
    v = my_model(X)[:,1:2]
    w = my_model(X)[:,2:3]
    p = my_model(X)[:,3:4]
```

**Reply by forxltk** [reply]
This is a problem about how to use Keras/TensorFlow. You should check relative docs.
